File: sdks/python/client/src/max_client/runtime/api_error.py
from __future__ import annotations

import json
from typing import NotRequired, TypedDict


class MaxErrorPayload(TypedDict, total=False):
    code: str
    message: str
    details: NotRequired[str | None]


class MaxApiError(Exception):
    def __init__(
        self,
        *,
        method: str,
        path: str,
        status: int,
        body_text: str,
        payload: MaxErrorPayload | None = None,
    ) -> None:
        message = payload.get("message") if payload else None
        super().__init__(message or f"{method} {path} failed with status {status}")
        self.method = method
        self.path = path
        self.status = status
        self.body_text = body_text
        self.payload = payload
        self.error_code = payload.get("code") if payload else None
        self.error_message = payload.get("message") if payload else None
        self.error_details = payload.get("details") if payload else None


def parse_error_payload(body_text: str) -> MaxErrorPayload | None:
    if not body_text.strip():
        return None

    try:
        parsed = json.loads(body_text)
    except json.JSONDecodeError:
        return None

    if not isinstance(parsed, dict):
        return None

    payload: MaxErrorPayload = {}
    if isinstance(parsed.get("code"), str):
        payload["code"] = parsed["code"]
    if isinstance(parsed.get("message"), str):
        payload["message"] = parsed["message"]
    if isinstance(parsed.get("details"), str) or parsed.get("details") is None:
        if "details" in parsed:
            payload["details"] = parsed["details"]

    return payload or None

File: sdks/typescript/client/runtime/api-error.ts
export interface MaxApiErrorInit {
  method: string;
  path: string;
  status: number;
  bodyText: string;
  payload?: MaxErrorPayload;
}

export interface MaxErrorPayload {
  code?: string;
  message?: string;
  details?: string | null;
}

export class MaxApiError extends Error {
  readonly method: string;
  readonly path: string;
  readonly status: number;
  readonly bodyText: string;
  readonly payload?: MaxErrorPayload;
  readonly errorCode?: string;
  readonly errorMessage?: string;
  readonly errorDetails?: string | null;

  constructor(init: MaxApiErrorInit) {
    super(init.payload?.message ?? `${init.method} ${init.path} failed with status ${init.status}`);
    this.name = "MaxApiError";
    this.method = init.method;
    this.path = init.path;
    this.status = init.status;
    this.bodyText = init.bodyText;
    this.payload = init.payload;
    this.errorCode = init.payload?.code;
    this.errorMessage = init.payload?.message;
    this.errorDetails = init.payload?.details;
  }
}

export function parseMaxErrorPayload(bodyText: string): MaxErrorPayload | undefined {
  if (!bodyText.trim()) {
    return undefined;
  }

  try {
    const parsed = JSON.parse(bodyText) as Record<string, unknown>;
    if (typeof parsed !== "object" || parsed === null) {
      return undefined;
    }

    const payload: MaxErrorPayload = {};
    if (typeof parsed.code === "string") {
      payload.code = parsed.code;
    }
    if (typeof parsed.message === "string") {
      payload.message = parsed.message;
    }
    if (typeof parsed.details === "string" || parsed.details === null) {
      payload.details = parsed.details as string | null;
    }

    return payload.code || payload.message || "details" in payload ? payload : undefined;
  } catch {
    return undefined;
  }
}
